Added 8-bit Quantization support to OneHot.

amitsrivastava78 opened the pull request:
This is to solve issue #27765.

Commit:
Added 8-bit Quantization support to OneHot.

This is to solve issue #27765.

gbaned commented:
@alanchiao Could you PTAL and approve.

amitsrivastava78 commented:
@alanchiao , could you please review the PR and provide feedback. This is kind of important PR for me.

Regards
Amit

alanchiao commented:
Looking at it now

amitsrivastava78 commented:
> Looking at it now

@alanchiao , thank you so much for your time on the PR, can you please provide your feedback for the same. I would like to see this PR though with your help  and would put in the required effort to do so.

Regards
Amit

alanchiao commented on tensorflow/lite/toco/graph_transformations/hardcode_min_max.cc:
```diff
@@ -63,6 +63,31 @@ bool HardcodeMinMaxForL2Normalization(Model* model, Operator* op) {
+  const auto& input_minmax = oninput_array.GetMinMax();
+  CHECK(!output_array.minmax);
+  auto& output_minmax = output_array.GetOrCreateMinMax();
+  output_minmax.min = input_minmax.min <= 0. ? 0. : input_minmax.min;
```
Within this particular context (quantization scheme), you want to hardcode the output min and max values so that the range covers all possible output values. For an example, tanh's output goes from -1 to 1, so the hardcoding approximates that as close as possible (-127/128 to 1.0).

In the case of tf.one_hot, the output goes from off_value to on_value, so you ideally want to approximately hardcode the min as off_value and the max as on_value. You don't want to clamp them between 0 and 1. 

Let me double check the following with someone before you do anything: on top of all of this, you may need the computed zero point to be an integer value. In tanh, this could be why the min is set to -127/128 instead of -1. 
With that for tanh, the scale is (1 - (-127/128))/255 = 1/128, where 1 is the max and -127/128 is the min. The zero point is 0 - (-127/128)/scale  = 127, an integer as desired. If the min was -1, you'd get a zero point of 127.5.

If the above is true, it'll complicate this, and it may be easier for us to start with only supporting one_hot when off_value = 0 and on_value = 1.

amitsrivastava78 replied:
@alanchiao , thanks spending time on the PR, and providing me the review comments. I will wait for you to confirm before i do anything as suggested by you.

Regards
Amit

alanchiao reviewed (commented):
General comment: to land this change, we need to consider how this fits in the entire graph or user experience. Suppose the on_value is 1.0 and the off_value is 0.0. We can exactly represent these values in uint8 (1.0 becomes 255 and 0.0 becomes 0.0). That's all good, but what will the user or graph then do with these values?

In the case of one_hot being an output, the user will end up seeing 0's and 255's as the output instead of the usual 0's and 1.0's. If we consider a problem such as image classification with multiple labels, there may be a 1.0 that indicates what label we think fits the image and the rest are 0's. Now the user needs to use 255 instead of 1.0 or ideally, we bring the 255 back to 1.0. In this case, it may not even make sense to quantize one hot.

I think it'd be good to look at the graph that zbl@ provided and see where the one hot fits in the graph and if they would run into the above problem

amitsrivastava78 commented:
@alanchiao , thanks for spending time on the PR and providing the feedback, the idea here is that if we want to convert and run the ner_model.pb to tflite we need to quantize one_hot operator, this is basically used for word embedding problems, and in this case for Chinese entity recognition,

I have checked the graph of the above model as well and the on_val and off_val are 1 and 0 respectively, the output of one_hot is directed to mat_mul which again will not have any problems with the quantized output which we have suggested in this PR.

I hope you are satisfied with the above explanation and would approve the PR.

Awaiting your reply.

Regards
Amit

amitsrivastava78 commented:
@alanchiao , i have updated the code with support for quantization from tools and updated kernel registry with updated version for one_hot kernel.


Regards
Amit

amitsrivastava78 commented:
@alanchiao , any update from your side, could you please review the code and provide your feedback

Regards
Amit

amitsrivastava78 commented:
@alanchiao , can you pls spend some time on the PR and provide your feedback.

Regards
Amit

mellanox-github commented:
Can one of the admins verify this patch?

gbaned commented:
@amitsrivastava78 Could you please resolve the conflicts? Thanks!

tensorflowbutler commented:
It has been 15 days with no activity and the `awaiting response` label was assigned. Is this PR still valid? Assigning the `stalled` label. Please comment to reassure me that this is still being worked on.
